qualification/Jenkinsfile

```diff
@@ -0,0 +1,147 @@
+ *
+ * Additionally the Jenkins server also needs access to a Mellanox NIC that
+ * supports ibverbs. The ibverbs drivers need to be passed into the Jenkins
+ * container using specific flags.
+ *
+ * For more information on the qualification framework, see katgpucbf/doc/qualification.rst
+ */
+
+pipeline {
+  agent {
+    dockerfile {
+
+      /* Use the Jenkins-specific stage of the Dockerfile as the image for
+       * testing. This provides the appropriate dependencies.
+       */
+      additionalBuildArgs '--target=jenkins'
+
+      /* The following argument needs to be specified in order for the container
+       * to launch correctly.
+       *
+       * --gpus=all: This argument passes the Nvidia driver and devices from the
+       * host to the container. It requires the NVIDIA Container Runtime to be
+       * installed on the host.
+       *
+       * -v /var/run/docker.sock:/var/run/docker.sock: makes the connection to
+       * the Docker server available inside the container (for building a Docker
+       * image).
+       *
+       * --network=host: The docker container needs access to the internet to
+       * install packages and an interface to connect to the correlator network.
+       * This command passes all the host interfaces to the container to be
+       * used as required.
+       *
+       * --device=/dev/infiniband/rdma_cm and --device=/dev/infiniband/uverbs0:
+       * These flags pass the drivers required for ibverbs to the container.
+       * NOTE: The driver is not always ubverbs0, sometimes it is ubverbs1 or
+       * 2 etc. The correct once needs to be specified.
+       */
+      args '--gpus=all -v /var/run/docker.sock:/var/run/docker.sock --network=host -u=root --ulimit=memlock=-1 --device=/dev/infiniband/rdma_cm  --device=/dev/infiniband/uverbs0'
+    }
+  }
+
+  options {
+    timeout(time: 1, unit: 'HOURS')
+  }
+
+  stages {
+    stage('Install Python packages') {
+      steps {
+        sh 'pip install -r requirements.txt -r requirements-dev.txt'
+      }
+    }
+
+    stage('Install katgpucbf package') {
+      steps {
+        sh 'pip install --no-deps ".[test]" && pip check'
+      }
+    }
+
+    stage('Parallel stage') {
+      parallel {
+        stage('Compile C++ tools') {
+          steps {
+            // Make and compile tools.
+            dir('src/tools') {
+              sh 'make clean'
+              sh 'make -j'
+            }
+          }
+        }
+
+        /* This stage ensures that all the python style guidelines checks pass.
+         * This will catch if someone has committed to the repo without
+         * installing the required pre-commit hooks, or has bypassed them.
+         */
+        stage('Run pre-commit checks') {
+          steps {
+            sh 'pre-commit install'
+            // no-commit-to-branch complains if we are on the main branch
+            sh 'SKIP=no-commit-to-branch pre-commit run --all-files'
+          }
+        }
+      }
+    }
```

Comment:
Shouldn't need to do any of this - it's already handled by the per-commit Jenkinsfile. However after installing qualification/requirements.txt you'll probably need to `sh 'pip install --no-deps ".[qualification]" && pip check'`.

Reply:
I added some missing modules to qualification/requirements.in, but this causing the building of the documentation to fail.

Reply:
Thats... odd. Documentation builds shouldn't touch the qualification directory. I can take a look later today after I'm top of admin.

Reply:
Thanks.

Reply:
@bmerry There is a merge conflict in the changes made to the qualification/requirements.txt file. Should I keep the changes I made to this file?

Reply:
What I'd suggest is:
1. Merge main into your branch, keeping the requirements.txt from `main` (which will bring all the version bumps from `main`).
2. Re-run pip-compile, which will add in the requirements you added to requirements.in.